File: enve/Services/Intelligence/BookIntelligenceModels.swift
import Foundation

enum BookIntelligenceScope: String, CaseIterable, Codable, Identifiable, Sendable {
    case previousChapter
    case currentChapterSoFar
    case lastTenMinutes
    case bookSoFar

    var id: String { rawValue }

    var title: String {
        switch self {
        case .previousChapter: return "Previous Chapter"
        case .currentChapterSoFar: return "This Chapter"
        case .lastTenMinutes: return "Last 10 Min"
        case .bookSoFar: return "Book So Far"
        }
    }

    func title(for mediaType: AppMediaType) -> String {
        switch mediaType {
        case .ebook:
            switch self {
            case .previousChapter: return "Previous Chapter"
            case .currentChapterSoFar: return "This Chapter"
            case .lastTenMinutes: return "Recent Pages"
            case .bookSoFar: return "Book So Far"
            }
        default:
            switch self {
            case .previousChapter: return "Chapter End"
            case .currentChapterSoFar: return "Recent Chapter"
            case .lastTenMinutes: return "Last 2 Min"
            case .bookSoFar: return "Recent Context"
            }
        }
    }

    var promptName: String {
        switch self {
        case .previousChapter: return "the previous chapter"
        case .currentChapterSoFar: return "the current chapter up to the current playback position"
        case .lastTenMinutes: return "the last ten minutes"
        case .bookSoFar: return "the book so far"
        }
    }

    func promptName(for mediaType: AppMediaType) -> String {
        switch mediaType {
        case .ebook:
            switch self {
            case .previousChapter: return "the previous chapter"
            case .currentChapterSoFar: return "the current chapter up to the current reading position"
            case .lastTenMinutes: return "the recent pages before the current reading position"
            case .bookSoFar: return "the book so far"
            }
        default:
            switch self {
            case .previousChapter: return "the end of the previous chapter"
            case .currentChapterSoFar: return "the recent part of the current chapter"
            case .lastTenMinutes: return "the last couple of minutes before the current playback position"
            case .bookSoFar: return "the recent local context near the current playback position"
            }
        }
    }
}

enum BookTranscriptStatus: String, Codable, Sendable {
    case missing
    case ready
    case stale
    case generating
    case failed
}

enum BookContextSource: String, Codable, Sendable {
    case audiobookTranscript
    case ebookText
}

struct TranscriptSourceFile: Codable, Equatable, Sendable {
    let path: String
    let byteCount: Int64
    let modificationTime: TimeInterval
}

struct TranscriptSourceFingerprint: Codable, Equatable, Sendable {
    let version: Int
    let localeIdentifier: String
    let duration: TimeInterval
    let files: [TranscriptSourceFile]
}

struct TranscriptSegment: Identifiable, Codable, Equatable, Sendable {
    let id: String
    let bookStableId: String
    let chapterId: String?
    let trackIndex: Int
    let startTime: TimeInterval
    let endTime: TimeInterval
    let text: String
    let confidence: Double?
    let isFinal: Bool

    var duration: TimeInterval {
        max(0, endTime - startTime)
    }
}

struct BookTranscriptManifest: Codable, Equatable, Sendable {
    let bookStableId: String
    var status: BookTranscriptStatus
    var localeIdentifier: String
    var createdAt: Date
    var updatedAt: Date
    var duration: TimeInterval
    var segmentCount: Int
    var sourceFingerprint: TranscriptSourceFingerprint?
    var failureMessage: String?
}

struct BookTranscript: Codable, Equatable, Sendable {
    var manifest: BookTranscriptManifest
    var segments: [TranscriptSegment]
}

struct EbookContextChunk: Identifiable, Codable, Equatable, Sendable {
    let id: String
    let bookStableId: String
    let title: String?
    let href: String?
    let index: Int
    let startProgress: Double
    let endProgress: Double
    let text: String
}

struct EbookContextManifest: Codable, Equatable, Sendable {
    let bookStableId: String
    var status: BookTranscriptStatus
    var createdAt: Date
    var updatedAt: Date
    var chunkCount: Int
    var failureMessage: String?
}

struct EbookContext: Codable, Equatable, Sendable {
    var manifest: EbookContextManifest
    var chunks: [EbookContextChunk]
}

struct BookContextResult: Sendable {
    let scope: BookIntelligenceScope
    let source: BookContextSource
    let range: ClosedRange<TimeInterval>
    let text: String
    let segmentCount: Int

    var isEmpty: Bool {
        text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

struct LibrarianMessage: Identifiable, Codable, Equatable, Sendable {
    enum Role: String, Codable, Sendable {
        case user
        case assistant
        case system
    }

    let id: UUID
    let role: Role
    let text: String
    let scope: BookIntelligenceScope?
    let createdAt: Date

    init(
        id: UUID = UUID(),
        role: Role,
        text: String,
        scope: BookIntelligenceScope? = nil,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.role = role
        self.text = text
        self.scope = scope
        self.createdAt = createdAt
    }
}

File: android/engine/src/main/java/com/enve/app/data/librarian/BookIntelligenceModels.kt
package com.enve.app.data.librarian

import com.enve.app.document.EbookSourceFormat
import com.enve.core.data.model.BookSource
import kotlinx.serialization.Serializable
import java.util.UUID

enum class BookIntelligenceScope {
    PREVIOUS_CHAPTER,
    CURRENT_CHAPTER_SO_FAR,
    RECENT_PAGES,
    BOOK_SO_FAR;

    val title: String
        get() = when (this) {
            PREVIOUS_CHAPTER -> "Previous Chapter"
            CURRENT_CHAPTER_SO_FAR -> "This Chapter"
            RECENT_PAGES -> "Recent Pages"
            BOOK_SO_FAR -> "Book So Far"
        }

    val promptName: String
        get() = when (this) {
            PREVIOUS_CHAPTER -> "the previous chapter"
            CURRENT_CHAPTER_SO_FAR -> "the current chapter up to the current reading position"
            RECENT_PAGES -> "the recent pages before the current reading position"
            BOOK_SO_FAR -> "the book so far"
        }
}

enum class BookContextStatus {
    MISSING,
    READY,
    GENERATING,
    FAILED,
}

enum class LibrarianEnginePreference {
    AUTOMATIC,
    REMOTE_SERVER,
    GEMINI_NANO,
    LITERT_LM,
    LOCAL_EXTRACTIVE;

    val title: String
        get() = when (this) {
            AUTOMATIC -> "Automatic"
            REMOTE_SERVER -> "Local Server"
            GEMINI_NANO -> "Gemini Nano"
            LITERT_LM -> "Local Model"
            LOCAL_EXTRACTIVE -> "Basic Local"
        }
}

enum class LibrarianEngineAvailability {
    AVAILABLE,
    DOWNLOADABLE,
    DOWNLOADING,
    MODEL_MISSING,
    UNAVAILABLE,
}

data class LibrarianEngineStatus(
    val preference: LibrarianEnginePreference,
    val availability: LibrarianEngineAvailability,
    val title: String,
    val detail: String,
    val isUsable: Boolean,
)

data class RecommendedLibrarianModel(
    val title: String,
    val detail: String,
    val sizeBytes: Long,
)

data class LibrarianModelDownloadState(
    val isActive: Boolean,
    val progress: Double?,
    val errorMessage: String?,
)

data class LibrarianAnswer(
    val text: String,
    val engineTitle: String,
)

@Serializable
data class LibrarianBookRef(
    val bookId: String,
    val sourceName: String,
    val connectionId: String?,
    val title: String,
    val author: String?,
    val formatName: String?,
    val currentProgress: Double,
) {
    val source: BookSource
        get() = runCatching { BookSource.valueOf(sourceName) }.getOrDefault(BookSource.GRIMMORY)

    val sourceFormat: EbookSourceFormat
        get() = EbookSourceFormat.fromServerType(formatName).let { format ->
            if (format == EbookSourceFormat.UNKNOWN) EbookSourceFormat.EPUB else format
        }

    val stableId: String
        get() = listOfNotNull(connectionId, sourceName, bookId).joinToString(":")

    val legacyStableId: String
        get() = "$sourceName:$bookId"
}

@Serializable
data class EbookContextManifest(
    val bookStableId: String,
    val status: BookContextStatus,
    val createdAtEpochMs: Long,
    val updatedAtEpochMs: Long,
    val chunkCount: Int,
    val failureMessage: String? = null,
)

@Serializable
data class EbookContextChunk(
    val id: String,
    val bookStableId: String,
    val title: String?,
    val href: String?,
    val index: Int,
    val startProgress: Double,
    val endProgress: Double,
    val text: String,
)

@Serializable
data class EbookContext(
    val manifest: EbookContextManifest,
    val chunks: List<EbookContextChunk>,
)

data class BookContextResult(
    val scope: BookIntelligenceScope,
    val rangeStart: Double,
    val rangeEnd: Double,
    val text: String,
    val chunkCount: Int,
) {
    val isEmpty: Boolean
        get() = text.isBlank()
}

@Serializable
data class LibrarianMessage(
    val id: String = UUID.randomUUID().toString(),
    val role: LibrarianMessageRole,
    val text: String,
    val scope: BookIntelligenceScope? = null,
    val engineTitle: String? = null,
    val createdAtEpochMs: Long = System.currentTimeMillis(),
)

@Serializable
enum class LibrarianMessageRole {
    USER,
    ASSISTANT,
}
